mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1896,6 +1857,157 @@ pub fn big_decimal_from_sat(satoshis: i64, decimals: u8) -> BigDecimal {
+    let standard_err = match Address::from_str(from) {
+        Ok(a) => return Ok(a),
+        Err(e) => e,
+    };
+
+    let cashaddress_err =
+        match Address::from_cashaddress(from, coin.checksum_type, coin.pub_addr_prefix, coin.p2sh_addr_prefix) {
+            Ok(a) => return Ok(a),
+            Err(e) => e,
+        };
+
+    ERR!(
+        "error on parse standard address: {:?}, error on parse cashaddress: {:?}",
+        standard_err,
+        cashaddress_err,
+    )
+}
+
+fn validate_payment<T>(
+    coin: T,
+    payment_tx: &[u8],
+    time_lock: u32,
+    first_pub0: &Public,
+    second_pub0: &Public,
+    priv_bn_hash: &[u8],
+    amount: BigDecimal,
+) -> Box<dyn Future<Item = (), Error = String> + Send>
+where
+    T: AsRef<UtxoCoinFields> + Send + Sync + 'static,
+{
+    let tx: UtxoTx = try_fus!(deserialize(payment_tx).map_err(|e| ERRL!("{:?}", e)));
+    let amount = try_fus!(sat_from_big_decimal(&amount, coin.as_ref().decimals));
+
+    let expected_redeem = payment_script(
+        time_lock,
+        priv_bn_hash,
+        &try_fus!(Public::from_slice(first_pub0)),
+        &try_fus!(Public::from_slice(second_pub0)),
+    );
+    let fut = async move {
+        let mut attempts = 0;
+        loop {
+            let tx_from_rpc = match coin
+                .as_ref()
+                .rpc_client
+                .get_transaction_bytes(tx.hash().reversed().into())
+                .compat()
+                .await
+            {
+                Ok(t) => t,
+                Err(e) => {
+                    if attempts > 2 {
+                        return ERR!(
+                            "Got error {:?} after 3 attempts of getting tx {:?} from RPC",
+                            e,
+                            tx.tx_hash()
+                        );
+                    };
+                    attempts += 1;
+                    log!("Error " [e] " getting the tx " [tx.tx_hash()] " from rpc");
+                    Timer::sleep(10.).await;
+                    continue;
+                },
+            };
+            if serialize(&tx).take() != tx_from_rpc.0 {
+                return ERR!(
+                    "Provided payment tx {:?} doesn't match tx data from rpc {:?}",
+                    tx,
+                    tx_from_rpc
+                );
+            }
+
+            let expected_output = TransactionOutput {
+                value: amount,
+                script_pubkey: Builder::build_p2sh(&dhash160(&expected_redeem)).into(),
+            };
+
+            if tx.outputs[0] != expected_output {
+                return ERR!(
+                    "Provided payment tx output doesn't match expected {:?} {:?}",
+                    tx.outputs[0],
+                    expected_output
+                );
+            }
+            return Ok(());
+        }
+    };
+    Box::new(fut.boxed().compat())
+}
+
+fn search_for_swap_tx_spend(
```

Comment:
Could you please refactor this to async?

Reply:
Refactored